src/transaction.ts

```diff
@@ -0,0 +1,359 @@
+   */
+  getChainId(): number {
+    return this._chainId
+  }
+
+  /**
+   * returns the sender's address
+   */
+  getSenderAddress(): Buffer {
+    if (this._from) {
+      return this._from
+    }
+    const pubkey = this.getSenderPublicKey()
+    this._from = publicToAddress(pubkey)
+    return this._from
+  }
+
+  /**
+   * returns the public key of the sender
+   */
+  getSenderPublicKey(): Buffer {
+    if (!this.verifySignature()) {
+      throw new Error('Invalid Signature')
+    }
+
+    // If the signature was verified successfully the _senderPubKey field is defined
+    return this._senderPubKey!
+  }
+
+  /**
+   * Determines if the signature is valid
+   */
+  verifySignature(): boolean {
+    const msgHash = this.hash(false)
+    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
+    if (this._common.gteHardfork('homestead') && new BN(this.s).cmp(N_DIV_2) === 1) {
+      return false
+    }
+
+    try {
+      const v = bufferToInt(this.v)
+      const useChainIdWhileRecoveringPubKey =
+        v >= this._chainId * 2 + 35 && this._common.gteHardfork('spuriousDragon')
+      this._senderPubKey = ecrecover(
+        msgHash,
+        v,
+        this.r,
+        this.s,
+        useChainIdWhileRecoveringPubKey ? this._chainId : undefined,
+      )
+    } catch (e) {
+      return false
+    }
+
+    return !!this._senderPubKey
+  }
+
+  /**
+   * sign a transaction with a given private key
+   * @param privateKey - Must be 32 bytes in length
+   */
+  sign(privateKey: Buffer) {
+    const msgHash = this.hash(false)
+    const sig = ecsign(msgHash, privateKey)
+    if (this._chainId > 0) {
+      sig.v += this._chainId * 2 + 8
+    }
+    Object.assign(this, sig)
+  }
+
+  /**
+   * The amount of gas paid for the data in this tx
+   */
+  getDataFee(): BN {
+    const data = this.raw[5]
+    const cost = new BN(0)
+    for (let i = 0; i < data.length; i++) {
+      data[i] === 0
+        ? cost.iaddn(this._common.param('gasPrices', 'txDataZero'))
+        : cost.iaddn(this._common.param('gasPrices', 'txDataNonZero'))
+    }
+    return cost
+  }
+
+  /**
+   * the minimum amount of gas the tx must have (DataFee + TxFee + Creation Fee)
+   */
+  getBaseFee(): BN {
+    const fee = this.getDataFee().iaddn(this._common.param('gasPrices', 'tx'))
+    if (this._common.gteHardfork('homestead') && this.toCreationAddress()) {
+      fee.iaddn(this._common.param('gasPrices', 'txCreation'))
+    }
+    return fee
+  }
+
+  /**
+   * the up front amount that an account must have for this transaction to be valid
+   */
+  getUpfrontCost(): BN {
+    return new BN(this.gasLimit).imul(new BN(this.gasPrice)).iadd(new BN(this.value))
+  }
+
+  /**
+   * Validates the signature and checks to see if it has enough gas.
+   */
+  validate(): boolean
+  validate(stringError: false): boolean
+  validate(stringError: true): string
+  validate(stringError: boolean = false): boolean | string {
```

Comment:
Is this last definition necessary for typescript? Don't the first three definitions cover all cases?

Reply:
This is something tricky, but really useful, about typescript. When you are overloading functions/methods, the last one is not exposed. It's just the implementation of the preceding overloads.

This may look weird initially, but it lets you expose very clean overloads. If the implementation were exposed, there will always be a weird overload.